lib/src/background_usecase.dart

```diff
@@ -0,0 +1,172 @@
+///    List<List<double>> result = List<List<double>>.generate(
+///        10, (i) => List<double>.generate(10, (j) => 0));
+///
+///    for (int i = 0; i < matMulParams.mat1.length; i++) {
+///      for (int j = 0; j < matMulParams.mat1.length; j++) {
+///        for (int k = 0; k < matMulParams.mat1.length; k++) {
+///          result[i][j] += matMulParams.mat1[i][k] * matMulParams.mat2[k][j];
+///        }
+///      }
+///    }
+///    params.port.send(BackgroundUseCaseMessage(data: result));
+///
+///  }
+///}
+///
+/// ```
+/// Just like a regular [UseCase], a parameter class is recommended for any [BackgroundUseCase].
+/// An example corresponding to the above example would be
+/// ```dart
+/// class MatMulUseCaseParams {
+///  List<List<double>> mat1;
+///  List<List<double>> mat2;
+///  MatMulUseCaseParams(this.mat1, this.mat2);
+///  MatMulUseCaseParams.random() {
+///    var size = 10;
+///    mat1 = List<List<double>>.generate(size,
+///        (i) => List<double>.generate(size, (j) => i.toDouble() * size + j));
+///
+///    mat2 = List<List<double>>.generate(size,
+///        (i) => List<double>.generate(size, (j) => i.toDouble() * size + j));
+///  }
+///}
+/// ```
+abstract class BackgroundUseCase<T, Params> extends UseCase<T, Params> {
+  BehaviorSubject<T> _subject;
+  BackgroundUseCaseState _state = BackgroundUseCaseState.idle;
+  Isolate _isolate;
+  final ReceivePort _receivePort;
+  static UseCaseTask _run;
+
+  BackgroundUseCase()
+      : _receivePort = ReceivePort(),
+        _subject = BehaviorSubject(),
+        super() {
+    _receivePort.listen(_handleMessage);
+  }
+
+  BackgroundUseCaseState get state => _state;
+  bool get isRunning => _state != BackgroundUseCaseState.idle;
+
+  /// Executes the usecase on a different isolate. Spawns [_isolate]
+  /// using the static method provided by [buildUseCaseTask] and listens
+  /// to a [BehaviorSubject] using the [observer] provided by the user.
+  /// All [Params] are sent to the [_isolate] through [BackgroundUseCaseParams].
+  @override
+  void execute(Observer<T> observer, [Params params]) async {
+    if (!isRunning) {
+      _state = BackgroundUseCaseState.loading;
+      _subject.listen(observer.onNext,
+          onError: observer.onError, onDone: observer.onComplete);
+      _run = buildUseCaseTask();
+      Isolate.spawn<BackgroundUseCaseParams>(_run,
+              BackgroundUseCaseParams(_receivePort.sendPort, params: params))
+          .then<void>((Isolate isolate) {
+        if (!isRunning) {
+          isolate.kill(priority: Isolate.immediate);
+        } else {
+          _state = BackgroundUseCaseState.calculating;
+          _isolate = isolate;
+        }
+      });
+    }
+  }
+
+  @override
+  @nonVirtual
+  Future<Observable<T>> buildUseCaseObservable(_) => null;
+
+  /// Provides a [UseCaseTask] to be executed on a different isolate.
+  /// Must be overridden.
+  UseCaseTask buildUseCaseTask();
```

Comment:
If must be overridden, shouldn't have `@override` annotation?

```suggestion
  @override
  UseCaseTask buildUseCaseTask();
```

Reply:
Hi,

The isolate only accepts static compile-time or top-level functions. Therefore, `_run = buildUseCaseTask` does not work since it is a non-static method. In addition, if it were changed to static, it would not be able to be overridden since static methods are not inherited.

This is why the user should define their own static method and simply return a reference to it.

The  `@override` annotation indicates that you are overriding a method, not that a method has to be overridden so it cannot be used here.

In order to use a static method, the `UseCaseTask` workaround has to remain.